### PR TITLE
[RLlib] Issue 15483: Wrong init states (should be non-zero if `ModelV2.get_initial_state` returns non-zero values).

### LINKED ISSUE #15483: [rllib] Incorrect initial state when using RNN or LSTM auto wrapping
### What is the problem?

*Ray version and other system information (Python version, TensorFlow version, OS):
ray: master

The `RecurrentNetwork.get_initial_state()` method is not being called between episodes during training. 
Instead the state_\*_\* keys are being filled with a default value of 0 in `SimpleListCollector.get_inference_input_dict`.
This breaks RNN models that do not use 0 as their initial state. 
I read the docs, checked issues here, and checked the disscusion forum but could not find anything indicating that custom initial states are no longer supported or how to set a ViewRequirement that calls get_initial_state().

I think there needs to be an `add_init_state` method added to `SampleCollector` that is called in the `all_agents_done` branch of `Sampler._process_observations`.

In the example below at t=0 state_in_0 and state_in_1 are set to all 0.1 and 0.5s respectively 

### Reproduction (REQUIRED)
```python
import argparse
import os

from ray.rllib.agents.ppo import PPOTrainer, DEFAULT_CONFIG
from ray.rllib.agents.trainer import merge_dicts
from ray.rllib.examples.env.stateless_cartpole import StatelessCartPole
from ray.rllib.examples.models.rnn_model import RNNModel, TorchRNNModel
from ray.rllib.agents.callbacks import DefaultCallbacks

parser = argparse.ArgumentParser()
parser.add_argument("--run", type=str, default="PPO")
parser.add_argument("--num-cpus", type=int, default=0)
parser.add_argument(
    "--framework", choices=["tf2", "tf", "tfe", "torch"], default="tf")
parser.add_argument("--as-test", action="store_true")
parser.add_argument("--use-prev-action", action="store_true")
parser.add_argument("--use-prev-reward", action="store_true")
parser.add_argument("--stop-iters", type=int, default=200)
parser.add_argument("--stop-timesteps", type=int, default=100000)
parser.add_argument("--stop-reward", type=float, default=150.0)



class PrintRNN(RNNModel):
    def get_initial_state(self):
        print("TF RNN: Get Initial State Called")
        s0, s1 =  super(PrintRNN, self).get_initial_state()
        return s0+0.1, s1+0.5

class PrintTorchRNN(TorchRNNModel):
    def get_initial_state(self):
        print("Torch RNN: Get Initial State Called")
        s0, s1 = super(PrintTorchRNN, self).get_initial_state()
        return s0+0.1, s1+0.5


class CheckInitState(DefaultCallbacks):
    def on_postprocess_trajectory(
            self, worker, episode,
            agent_id, policy_id,
            policies, postprocessed_batch,
            original_batches, **kwargs):
        stored_state = postprocessed_batch["state_in_0"][0], postprocessed_batch["state_in_1"][0]
        init_state = policies["default_policy"].model.get_initial_state()
        if args.framework == "torch":
            assert stored_state==init_state, \
                f"actual: {stored_state} \n" \
                f"expected: {init_state}"
        else:
            assert  (stored_state[0]==init_state[0]).all() and (stored_state[1]==init_state[1]).all(),\
                f"actual: {stored_state} \n" \
                f"expected: {init_state}"

if __name__ == "__main__":
    import ray
    from ray import tune

    args = parser.parse_args()

    ray.init(local_mode=True, num_cpus=args.num_cpus or None)

    configs = {
        "PPO": {
            "num_workers": 0,
            "num_sgd_iter": 5,
            "model": {
                "vf_share_layers": True,
            },
            "vf_loss_coeff": 0.0001,
        },
    }

    config = dict(
        configs[args.run],
        **{
            "env": StatelessCartPole,
            # Use GPUs iff `RLLIB_NUM_GPUS` env var set to > 0.
            "num_gpus": int(os.environ.get("RLLIB_NUM_GPUS", "0")),
            "model": {
                "custom_model": PrintTorchRNN if args.framework == "torch" else PrintRNN,
                "custom_model_config": {"lstm_state_size": 10} if args.framework == "torch"
                                        else {"cell_size": 10},
            },
            "framework": args.framework,
            # Run with tracing enabled for tfe/tf2.
            "eager_tracing": args.framework in ["tfe", "tf2"],
            "callbacks": CheckInitState,
        })


    config = merge_dicts(DEFAULT_CONFIG, config)
    trainer = PPOTrainer(config=config)
    for i in range(10):
        trainer.train()

    ray.shutdown()
```

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by sven1977**
Issue 15483: Wrong init states (should be non-zero if `ModelV2.get_initial_state` returns non-zero values).

Currently, if ModelV2.get_initial_state returns a non-zero tensor, RLlib ignores this and initializes the states in the SampleCollector's pre-buffer with 0s either way. This is incorrect.

#15483 
## Why are these changes needed?

Closes #15483 
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by sven1977** [comment]
@mvindiola1 , if you get a chance, could you confirm that this solves your problem with the init states being able to have arbitrary values?
I think all CI tests should be pass now (waiting for confirmation on this).

File: rllib/policy/policy.py [hunk 1]
```diff
@@ -789,9 +789,15 @@ def _update_model_view_requirements_from_init_state(self):
                 obj.get_initial_state = lambda: []
                 if "state_in_0" in view_reqs:
                     self.is_recurrent = lambda: True
+
         for i, state in enumerate(init_state):
-            space = Box(-1.0, 1.0, shape=state.shape) if \
-                hasattr(state, "shape") else state
+            fw = np if isinstance(state, np.ndarray) else torch if \
```
Comment (sven1977) [inline rllib/policy/policy.py:794]:
```suggestion
            # Allow `state` to be either a Space (use zeros as initial values)
            # or any value (e.g. a dict or a non-zero tensor).
            fw = np if isinstance(state, np.ndarray) else torch if \
```

**Comment by sven1977** [comment]
Awesome @mvindiola1 ! This fix should go into the upcoming release 1.4, I hope.

**Comment by sven1977** [comment]
and thanks for confirming.

**Review by richardliaw — APPROVED** [review]
(rubber stamp)

**Comment by mg64ve** [comment]
@mvindiola1 could you please share your updated code? I would like to make a test. Thanks.

**Comment by mg64ve** [comment]
with the following code:

```
import argparse
import os

from ray.rllib.agents.ppo import PPOTrainer, DEFAULT_CONFIG
from ray.rllib.agents.trainer import merge_dicts
from ray.rllib.examples.env.stateless_cartpole import StatelessCartPole
from ray.rllib.examples.models.rnn_model import RNNModel, TorchRNNModel
from ray.rllib.agents.callbacks import DefaultCallbacks

parser = argparse.ArgumentParser()
parser.add_argument("--run", type=str, default="PPO")
parser.add_argument("--num-cpus", type=int, default=0)
parser.add_argument("--framework", choices=["tf2", "tf", "tfe", "torch"], default="tf")
parser.add_argument("--as-test", action="store_true")
parser.add_argument("--use-prev-action", action="store_true")
parser.add_argument("--use-prev-reward", action="store_true")
parser.add_argument("--stop-iters", type=int, default=200)
parser.add_argument("--stop-timesteps", type=int, default=100000)
parser.add_argument("--stop-reward", type=float, default=150.0)



class PrintRNN(RNNModel):
    def get_initial_state(self):
        print("TF RNN: Get Initial State Called")
        s0, s1 =  super(PrintRNN, self).get_initial_state()
        return s0+0.1, s1+0.5

class PrintTorchRNN(TorchRNNModel):
    def get_initial_state(self):
        print("Torch RNN: Get Initial State Called")
        s0, s1 = super(PrintTorchRNN, self).get_initial_state()
        return s0+0.1, s1+0.5


class CheckInitState(DefaultCallbacks):
    def on_postprocess_trajectory(
            self, worker, episode,
            agent_id, policy_id,
            policies, postprocessed_batch,
            original_batches, **kwargs):
        if len(episode.user_data["pole_angles"]) == 0:
            stored_state = postprocessed_batch["state_in_0"][0], postprocessed_batch["state_in_1"][0]
            init_state = policies["default_policy"].model.get_initial_state()
            if args.framework == "torch":
                assert stored_state==init_state, \
                    f"actual: {stored_state} \n" \
                    f"expected: {init_state}"
            else:
                assert  (stored_state[0]==init_state[0]).all() and (stored_state[1]==init_state[1]).all(),\
                    f"actual: {stored_state} \n" \
                    f"expected: {init_state}"

if __name__ == "__main__":
    import ray
    from ray import tune

    args = parser.parse_args()

    ray.init(local_mode=True, num_cpus=args.num_cpus or None)

    configs = {
        "PPO": {
            "num_workers": 0,
            "num_sgd_iter": 5,
            "model": {
                "vf_share_layers": True,
            },
            "vf_loss_coeff": 0.0001,
        },
    }

    config = dict(
        configs[args.run],
        **{
            "env": StatelessCartPole,
            # Use GPUs iff `RLLIB_NUM_GPUS` env var set to > 0.
            "num_gpus": int(os.environ.get("RLLIB_NUM_GPUS", "0")),
            "model": {
                "custom_model": PrintTorchRNN if args.framework == "torch" else PrintRNN,
                "custom_model_config": {"lstm_state_size": 10} if args.framework == "torch"
                                        else {"cell_size": 10},
            },
            "framework": args.framework,
            # Run with tracing enabled for tfe/tf2.
            "eager_tracing": args.framework in ["tfe", "tf2"],
            "callbacks": CheckInitState,
        })


    config = merge_dicts(DEFAULT_CONFIG, config)
    trainer = PPOTrainer(config=config)
    for i in range(10):
        trainer.train()

    ray.shutdown()
```

I am getting the following error for torch:

```
/home/condauser/.local/lib/python3.8/site-packages/ray/rllib/policy/sample_batch.py:105: FutureWarning: The input object of type 'Tensor' is an array-like implementing one of the corresponding protocols (`__array__`, `__array_interface__` or `__array_struct__`); but not a sequence (or 0-D). In the future, this object will be coerced as if it was first converted using `np.array(obj)`. To retain the old behaviour, you have to either modify the type 'Tensor', or assign to an empty array created with `np.empty(correct_shape, dtype=object)`.
  self[k] = np.array(v)
/home/condauser/.local/lib/python3.8/site-packages/ray/rllib/policy/sample_batch.py:105: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
  self[k] = np.array(v)
Traceback (most recent call last):
  File "check_state.py", line 92, in <module>
    trainer = PPOTrainer(config=config)
  File "/home/condauser/.local/lib/python3.8/site-packages/ray/rllib/agents/trainer_template.py", line 123, in __init__
    Trainer.__init__(self, config, env, logger_creator)
  File "/home/condauser/.local/lib/python3.8/site-packages/ray/rllib/agents/trainer.py", line 562, in __init__
    super().__init__(config, logger_creator)
  File "/home/condauser/.local/lib/python3.8/site-packages/ray/tune/trainable.py", line 100, in __init__
    self.setup(copy.deepcopy(self.config))
  File "/home/condauser/.local/lib/python3.8/site-packages/ray/rllib/agents/trainer.py", line 722, in setup
    self._init(self.config, self.env_creator)
  File "/home/condauser/.local/lib/python3.8/site-packages/ray/rllib/agents/trainer_template.py", line 147, in _init
    self.workers = self._make_workers(
  File "/home/condauser/.local/lib/python3.8/site-packages/ray/rllib/agents/trainer.py", line 804, in _make_workers
    return WorkerSet(
  File "/home/condauser/.local/lib/python3.8/site-packages/ray/rllib/evaluation/worker_set.py", line 92, in __init__
    self._local_worker = self._make_worker(
  File "/home/condauser/.local/lib/python3.8/site-packages/ray/rllib/evaluation/worker_set.py", line 368, in _make_worker
    worker = cls(
  File "/home/condauser/.local/lib/python3.8/site-packages/ray/rllib/evaluation/rollout_worker.py", line 545, in __init__
    self.policy_map, self.preprocessors = self._build_policy_map(
  File "/home/condauser/.local/lib/python3.8/site-packages/ray/rllib/evaluation/rollout_worker.py", line 1323, in _build_policy_map
    policy_map[name] = cls(obs_space, act_space, merged_conf)
  File "/home/condauser/.local/lib/python3.8/site-packages/ray/rllib/policy/policy_template.py", line 279, in __init__
    self._initialize_loss_from_dummy_batch(
  File "/home/condauser/.local/lib/python3.8/site-packages/ray/rllib/policy/policy.py", line 625, in _initialize_loss_from_dummy_batch
    self.compute_actions_from_input_dict(
  File "/home/condauser/.local/lib/python3.8/site-packages/ray/rllib/policy/torch_policy.py", line 262, in compute_actions_from_input_dict
    return self._compute_action_helper(input_dict, state_batches,
  File "/home/condauser/.local/lib/python3.8/site-packages/ray/rllib/utils/threading.py", line 21, in wrapper
    return func(self, *a, **k)
  File "/home/condauser/.local/lib/python3.8/site-packages/ray/rllib/policy/torch_policy.py", line 326, in _compute_action_helper
    dist_inputs, state_out = self.model(input_dict, state_batches,
  File "/home/condauser/.local/lib/python3.8/site-packages/ray/rllib/models/modelv2.py", line 234, in __call__
    res = self.forward(restored, state or [], seq_lens)
  File "/home/condauser/.local/lib/python3.8/site-packages/ray/rllib/models/torch/recurrent_net.py", line 83, in forward
    output, new_state = self.forward_rnn(inputs, state, seq_lens)
  File "/home/condauser/.local/lib/python3.8/site-packages/ray/rllib/examples/models/rnn_model.py", line 133, in forward_rnn
    x, [torch.unsqueeze(state[0], 0),
TypeError: unsqueeze(): argument 'input' (position 1) must be Tensor, not numpy.ndarray
```

**Pull request merged**